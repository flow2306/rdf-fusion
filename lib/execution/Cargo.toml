[package]
name = "rdf-fusion-execution"
version.workspace = true
authors.workspace = true
license.workspace = true
edition.workspace = true
rust-version.workspace = true
description = """
RDF Fusion Execution Engine
"""

[dependencies]
async-trait.workspace = true
datafusion.workspace = true
itertools.workspace = true
rdf-fusion-extensions.workspace = true
rdf-fusion-encoding.workspace = true
rdf-fusion-functions.workspace = true
rdf-fusion-model.workspace = true
futures.workspace = true
rdf-fusion-logical.workspace = true
rdf-fusion-physical.workspace = true
oxrdfio.workspace = true
spargebra.workspace = true
sparesults.workspace = true
thiserror.workspace = true
tokio.workspace = true

[lints]
workspace = true
