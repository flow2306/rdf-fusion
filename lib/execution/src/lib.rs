#![doc(test(attr(deny(warnings))))]
#![doc(
    html_favicon_url = "https://raw.githubusercontent.com/tobixdev/rdf-fusion/main/misc/logo/logo.png"
)]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/tobixdev/rdf-fusion/main/misc/logo/logo.png"
)]

//! This crate defines the execution engine of RDF Fusion.
//!
//! # RDF Fusion Context
//!
//! Similar to DataFusion’s [`SessionContext`](datafusion::prelude::SessionContext),
//! RDF Fusion provides an [`RdfFusionContext`].
//! This context manages the state of the RDF Fusion engine, including the registered RDF term
//! encodings.
//! For more details, see the [`RdfFusionContext`] documentation.
//!
//! # Executing SPARQL Queries
//!
//! DataFusion operates on the concept of a *logical plan*.
//! While DataFusion includes an SQL front-end, it does not provide a SPARQL front-end.
//! To execute SPARQL queries, we convert them into logical plans through the following pipeline:
//!
//! ```text
//! Query String -> SPARQL Algebra (Oxigraph) -> Logical Plan
//! ```
//!
//! The query string is first parsed by Oxigraph’s SPARQL parser, which produces a SPARQL algebra
//! graph pattern. This graph pattern already includes high-level concepts such as `Join`, `Filter`,
//! and `Projection`.
//!
//! We then convert this algebra pattern into a DataFusion logical plan through a rewriting step.
//! The resulting logical plan is executed by DataFusion.

extern crate core;

mod engine;
mod planner;
pub mod results;
pub mod sparql;

pub use engine::RdfFusionContext;
