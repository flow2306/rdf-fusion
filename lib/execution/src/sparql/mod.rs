//! [SPARQL](https://www.w3.org/TR/sparql11-overview/) implementation.

mod algebra;
pub mod error;
mod eval;
mod explanation;
mod optimizer;
mod rewriting;

pub use crate::sparql::algebra::{Query, QueryDataset, Update};
pub use crate::sparql::explanation::QueryExplanation;
pub use eval::evaluate_query;
pub use optimizer::{create_optimizer_rules, create_pyhsical_optimizer_rules};
pub use rdf_fusion_model::{Variable, VariableNameParseError};
pub use spargebra::SparqlSyntaxError;

/// Defines how many optimizations the query optimizer should apply.
///
/// Currently, the default value is [OptimizationLevel::Full], as we are still searching for a
/// subset that performs well on many queries. Once this subset has been identified, the default
/// value will be [OptimizationLevel::Default].
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum OptimizationLevel {
    /// No optimizations, except rewrites that are necessary for a working query.
    None,
    /// A balanced default optimization level. Suitable for simple queries or those handling modest
    /// data volumes.
    Default,
    /// Runs all optimizations. Ideal for complex queries or those processing large datasets.
    #[default]
    Full,
}

/// Options for SPARQL query evaluation.
#[derive(Clone, Default)]
pub struct QueryOptions {
    /// The defined optimization level
    pub optimization_level: OptimizationLevel,
}

/// Options for SPARQL update evaluation.
#[derive(Clone, Default)]
pub struct UpdateOptions;

impl From<QueryOptions> for UpdateOptions {
    #[inline]
    fn from(_query_options: QueryOptions) -> Self {
        Self {}
    }
}
