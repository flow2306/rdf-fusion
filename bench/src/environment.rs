use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use datafusion::execution::runtime_env::{RuntimeEnv, RuntimeEnvBuilder};
use datafusion::prelude::SessionConfig;
use rdf_fusion::store::Store;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the results dir.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Creates a new [RdfFusionBenchContext] used in the criterion benchmarks.
    pub fn new_for_criterion(data_dir: PathBuf, target_partitions: usize) -> Self {
        Self {
            options: BenchmarkingOptions {
                verbose_results: false,
                target_partitions: Some(target_partitions),
                memory_size: None,
            },
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(PathBuf::from("/temp")),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    pub fn create_store(&self) -> Store {
        let mut config = SessionConfig::new()
            .with_batch_size(8192)
            .with_target_partitions(self.options.target_partitions.unwrap_or(1));

        let options = config.options_mut();
        options.optimizer.enable_dynamic_filter_pushdown = true;

        let runtime_enc = match self.options.memory_size {
            None => Arc::new(RuntimeEnv::default()),
            Some(memory_size) => RuntimeEnvBuilder::new()
                .with_memory_limit(memory_size * 1024 * 1024, 1f64)
                .build_arc()
                .expect("Only setting memory limit"),
        };
        Store::new_with_datafusion_config(config, runtime_enc)
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    #[allow(clippy::unwrap_used, reason = "Mutex poisoning")]
    pub fn push_dir(
        &self,
        data_dir_name: &str,
        results_dir_name: &str,
    ) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(data_dir_name);
        results_dir.push(results_dir_name);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(
            &benchmark_name.data_dir_name(),
            &benchmark_name.results_dir_name(),
        )?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
