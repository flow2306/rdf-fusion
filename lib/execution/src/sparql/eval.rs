use crate::RdfFusionContext;
use crate::results::{QueryResults, QuerySolutionStream, QueryTripleStream};
use crate::sparql::error::QueryEvaluationError;
use crate::sparql::optimizer::{create_optimizer_rules, create_pyhsical_optimizer_rules};
use crate::sparql::rewriting::GraphPatternRewriter;
use crate::sparql::{Query, QueryDataset, QueryExplanation, QueryOptions};
use datafusion::arrow::datatypes::Schema;
use datafusion::common::instant::Instant;
use datafusion::execution::{SessionState, SessionStateBuilder};
use datafusion::physical_plan::{ExecutionPlan, execute_stream};
use futures::StreamExt;
use itertools::izip;
use rdf_fusion_logical::RdfFusionLogicalPlanBuilderContext;
use rdf_fusion_model::Iri;
use rdf_fusion_model::Variable;
use spargebra::algebra::GraphPattern;
use spargebra::term::TriplePattern;
use std::sync::Arc;

/// Evaluates a SPARQL query and returns the results along with execution information.
///
/// Most users should refrain from directly using this function, as there are higher-level
/// abstractions that provide APIs for querying.
pub async fn evaluate_query(
    ctx: &RdfFusionContext,
    builder_context: RdfFusionLogicalPlanBuilderContext,
    query: &Query,
    options: QueryOptions,
) -> Result<(QueryResults, QueryExplanation), QueryEvaluationError> {
    let session_state = SessionStateBuilder::from(ctx.session_context().state())
        .with_optimizer_rules(create_optimizer_rules(
            ctx.create_view(),
            options.optimization_level,
        ))
        .with_physical_optimizer_rules(create_pyhsical_optimizer_rules(
            options.optimization_level,
        ))
        .build();

    match &query.inner {
        spargebra::Query::Select {
            pattern, base_iri, ..
        } => {
            let (stream, explanation) = Box::pin(graph_pattern_to_stream(
                session_state,
                builder_context,
                query,
                pattern,
                base_iri,
            ))
            .await?;
            Ok((QueryResults::Solutions(stream), explanation))
        }
        spargebra::Query::Construct {
            template,
            pattern,
            base_iri,
            ..
        } => {
            let (stream, explanation) = Box::pin(graph_pattern_to_stream(
                session_state,
                builder_context,
                query,
                pattern,
                base_iri,
            ))
            .await?;
            Ok((
                QueryResults::Graph(QueryTripleStream::new(template.clone(), stream)),
                explanation,
            ))
        }
        spargebra::Query::Ask {
            pattern, base_iri, ..
        } => {
            let (mut stream, explanation) = Box::pin(graph_pattern_to_stream(
                session_state,
                builder_context,
                query,
                pattern,
                base_iri,
            ))
            .await?;
            let count = stream.next().await;
            Ok((QueryResults::Boolean(count.is_some()), explanation))
        }
        spargebra::Query::Describe {
            pattern, base_iri, ..
        } => {
            // TODO: Research what a good DESCRIBE implementation would look like.

            let mut vars = Vec::new();
            pattern.on_in_scope_variable(|v| vars.push(v.clone()));
            let rdf_types = vars
                .iter()
                .map(|v| Variable::new(format!("{}__type", v.as_str())).unwrap())
                .collect::<Vec<_>>();

            let describe_pattern = izip!(vars, rdf_types.iter())
                .map(|(variable, rdf_type)| {
                    vec![TriplePattern {
                        subject: variable.clone().into(),
                        predicate: rdf_fusion_model::vocab::rdf::TYPE.into_owned().into(),
                        object: rdf_type.clone().into(),
                    }]
                    .into_iter()
                })
                .flatten()
                .collect::<Vec<_>>();

            // Compute the label / comment results
            let pattern = GraphPattern::Join {
                left: Box::new(pattern.clone()),
                right: Box::new(GraphPattern::Bgp {
                    patterns: describe_pattern.clone(),
                }),
            };
            let (stream, explanation) = Box::pin(graph_pattern_to_stream(
                session_state,
                builder_context,
                query,
                &pattern,
                base_iri,
            ))
            .await?;

            Ok((
                QueryResults::Graph(QueryTripleStream::new(describe_pattern, stream)),
                explanation,
            ))
        }
    }
}

/// Converts a SPARQL graph pattern to a stream of query solutions.
async fn graph_pattern_to_stream(
    state: SessionState,
    builder_context: RdfFusionLogicalPlanBuilderContext,
    query: &Query,
    pattern: &GraphPattern,
    base_iri: &Option<Iri<String>>,
) -> Result<(QuerySolutionStream, QueryExplanation), QueryEvaluationError> {
    let task = state.task_ctx();

    let (execution_plan, explanation) =
        create_execution_plan(state, builder_context, &query.dataset, pattern, base_iri)
            .await?;
    let variables = create_variables(&execution_plan.schema());

    let batch_record_stream = execute_stream(execution_plan, task)?;
    let stream = QuerySolutionStream::try_new(variables, batch_record_stream)?;
    Ok((stream, explanation))
}

/// Creates a physical execution plan from a SPARQL graph pattern, doing further processing on the
/// resulting query plan (e.g., optimization).
async fn create_execution_plan(
    state: SessionState,
    builder_context: RdfFusionLogicalPlanBuilderContext,
    dataset: &QueryDataset,
    pattern: &GraphPattern,
    base_iri: &Option<Iri<String>>,
) -> Result<(Arc<dyn ExecutionPlan>, QueryExplanation), QueryEvaluationError> {
    let planning_time_start = Instant::now();
    let logical_plan =
        GraphPatternRewriter::new(builder_context, dataset.clone(), base_iri.clone())
            .rewrite(pattern)
            .map_err(|e| e.context("Cannot rewrite SPARQL query"))?;
    let optimized_plan = state.optimize(&logical_plan)?;
    let physical_plan = state
        .query_planner()
        .create_physical_plan(&optimized_plan, &state)
        .await?;
    let planning_time = planning_time_start.elapsed();

    let explanation = QueryExplanation {
        planning_time,
        initial_logical_plan: logical_plan,
        optimized_logical_plan: optimized_plan,
        execution_plan: Arc::clone(&physical_plan),
    };
    Ok((Arc::clone(&physical_plan), explanation))
}

#[allow(clippy::expect_used)]
fn create_variables(schema: &Schema) -> Arc<[Variable]> {
    schema
        .fields()
        .iter()
        .map(|f| Variable::new(f.name()).expect("Variables already checked."))
        .collect::<Vec<_>>()
        .into()
}
