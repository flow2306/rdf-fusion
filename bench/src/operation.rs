use crate::runs::BenchmarkRun;
use futures::StreamExt;
use rdf_fusion::execution::results::QueryResults;
use rdf_fusion::execution::sparql::{Query, QueryExplanation, QueryOptions};
use rdf_fusion::store::Store;

#[derive(Clone)]
pub enum SparqlRawOperation<QueryName> {
    Query(QueryName, String),
}

impl<QueryName: Clone> SparqlRawOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlRawOperation::Query(name, _) => name.clone(),
        }
    }

    pub fn text(&self) -> &str {
        match self {
            SparqlRawOperation::Query(_, text) => text.as_ref(),
        }
    }

    pub fn parse(&self) -> anyhow::Result<SparqlOperation<QueryName>> {
        match self {
            SparqlRawOperation::Query(query_name, query) => {
                let query = query.parse()?;
                Ok(SparqlOperation::Query(query_name.clone(), query))
            }
        }
    }
}

#[derive(Clone)]
pub enum SparqlOperation<QueryName> {
    Query(QueryName, Query),
}

impl<QueryName> SparqlOperation<QueryName> {
    pub fn query(&self) -> &Query {
        match self {
            SparqlOperation::Query(_, query) => query,
        }
    }

    pub async fn run(
        &self,
        store: &Store,
    ) -> anyhow::Result<(BenchmarkRun, QueryExplanation, usize)> {
        let start = datafusion::common::instant::Instant::now();

        let mut num_results = 0;
        let options = QueryOptions::default();
        let explanation = match &self {
            SparqlOperation::Query(_, q) => {
                let (result, explanation) =
                    store.explain_query_opt(q.clone(), options.clone()).await?;
                match result {
                    QueryResults::Boolean(_) => (),
                    QueryResults::Solutions(s) => {
                        let mut stream = s.into_record_batch_stream()?;
                        while let Some(s) = stream.next().await {
                            num_results += s?.num_rows();
                        }
                    }
                    QueryResults::Graph(mut g) => {
                        while let Some(t) = g.next().await {
                            t?;
                            num_results += 1;
                        }
                    }
                }
                explanation
            }
        };

        let duration = start.elapsed();
        Ok((BenchmarkRun { duration }, explanation, num_results))
    }
}

impl<QueryName: Clone> SparqlOperation<QueryName> {
    pub fn query_name(&self) -> QueryName {
        match self {
            SparqlOperation::Query(name, _) => name.clone(),
        }
    }
}
