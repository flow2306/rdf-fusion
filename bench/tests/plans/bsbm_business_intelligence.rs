use crate::plans::run_plan_assertions;
use anyhow::Context;
use datafusion::physical_plan::displayable;
use insta::assert_snapshot;
use rdf_fusion::execution::sparql::{QueryExplanation, QueryOptions};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::bsbm::{
    BsbmBenchmark, BsbmBusinessIntelligenceQueryName, BusinessIntelligenceUseCase,
    NumProducts,
};
use rdf_fusion_bench::environment::{BenchmarkContext, RdfFusionBenchContext};
use rdf_fusion_bench::operation::SparqlRawOperation;
use std::path::PathBuf;

#[tokio::test]
pub async fn optimized_logical_plan_bsbm_business_intelligence() {
    for_all_explanations(|name, explanation| {
        assert_snapshot!(
            format!("{name} (Optimized)"),
            &explanation.optimized_logical_plan.to_string()
        )
    })
    .await;
}

#[tokio::test]
pub async fn execution_plan_bsbm_business_intelligence() {
    for_all_explanations(|name, explanation| {
        let string = displayable(explanation.execution_plan.as_ref())
            .indent(false)
            .to_string();
        assert_snapshot!(format!("{name} (Execution Plan)"), &string)
    })
    .await;
}

async fn for_all_explanations(assertion: impl Fn(String, QueryExplanation) -> ()) {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"), 1);

    // Load the benchmark data and set max query count to one.
    let benchmark =
        BsbmBenchmark::<BusinessIntelligenceUseCase>::try_new(NumProducts::N1_000, None)
            .unwrap();
    let benchmark_name = benchmark.name();
    let benchmark_context = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store = benchmark.prepare_store(&benchmark_context).await.unwrap();
    for query_name in BsbmBusinessIntelligenceQueryName::list_queries() {
        let benchmark_name = format!("BSBM Business Intelligence - {query_name}");
        let query =
            get_query_to_execute(benchmark.clone(), &benchmark_context, query_name);

        let (_, explanation) = store
            .explain_query_opt(query.text(), QueryOptions::default())
            .await
            .unwrap();

        run_plan_assertions(|| assertion(benchmark_name, explanation));
    }
}

fn get_query_to_execute(
    benchmark: BsbmBenchmark<BusinessIntelligenceUseCase>,
    benchmark_context: &BenchmarkContext,
    query_name: BsbmBusinessIntelligenceQueryName,
) -> SparqlRawOperation<BsbmBusinessIntelligenceQueryName> {
    benchmark
        .list_raw_operations(&benchmark_context)
        .context("Could not list raw operations for BSBM Business Intelligence benchmark. Have you prepared a bsbm-1000 dataset?")
        .unwrap()
        .into_iter()
        .find(|q| q.query_name() == query_name)
        .unwrap()
}
