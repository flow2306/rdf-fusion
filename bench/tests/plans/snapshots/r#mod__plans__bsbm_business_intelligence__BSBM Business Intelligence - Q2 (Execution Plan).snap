---
source: bench/tests/plans/bsbm_business_intelligence.rs
expression: "&string"
---
ProjectionExec: expr=[ENC_PT(otherProduct@0) as otherProduct, ENC_PT(sameFeatures@1) as sameFeatures]
  SortExec: TopK(fetch=10), expr=[ENC_SORT(sameFeatures@1) DESC, ENC_SORT(ENC_PT(otherProduct@0)) ASC], preserve_partitioning=[false]
    CoalesceBatchesExec: target_batch_size=8192
      HashJoinExec: mode=CollectLeft, join_type=Inner, on=[(otherProduct@0, otherProduct@0)], projection=[otherProduct@0, sameFeatures@2]
        CoalesceBatchesExec: target_batch_size=8192
          FilterExec: otherProduct@0 != <object id>
            DataSourceExec: [GPOS] subject=?otherProduct, predicate=<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>, object=<http://www4.wiwiss.fu-berlin.de/bizer/bsbm/v01/vocabulary/Product>
        ProjectionExec: expr=[otherProduct@0 as otherProduct, INT64_AS_TERM(<uuid>@1) as sameFeatures]
          CoalesceBatchesExec: target_batch_size=8192
            FilterExec: otherProduct@0 != <object id>
              AggregateExec: mode=Single, gby=[otherProduct@0 as otherProduct], aggr=[<uuid>]
                NestedLoopJoinExec: join_type=Inner, filter=EBV(EQ(join_proj_push_down_1@0, join_proj_push_down_2@1)), projection=[otherProduct@2, otherFeature@3]
                  ProjectionExec: expr=[feature@0 as feature, ENC_TV(feature@0) as join_proj_push_down_1]
                    CoalesceBatchesExec: target_batch_size=8192
                      DataSourceExec: [GSPO] subject=<http://www4.wiwiss.fu-berlin.de/bizer/bsbm/v01/instances/dataFromProducer13/Product636>, predicate=<http://www4.wiwiss.fu-berlin.de/bizer/bsbm/v01/vocabulary/productFeature>, object=?feature
                  ProjectionExec: expr=[otherProduct@0 as otherProduct, otherFeature@1 as otherFeature, ENC_TV(otherFeature@1) as join_proj_push_down_2]
                    CoalesceBatchesExec: target_batch_size=8192
                      DataSourceExec: [GPOS] subject=?otherProduct, predicate=<http://www4.wiwiss.fu-berlin.de/bizer/bsbm/v01/vocabulary/productFeature>, object=?otherFeature
